matrix/common/s3_zarr_store.py

```diff
@@ -0,0 +1,188 @@
+import os
+import math
+import json
+
+import s3fs
+import zarr
+from pandas import DataFrame
+import numpy
+
+
+from matrix.common.dynamo_handler import DynamoHandler
+from matrix.common.dynamo_handler import DynamoTable
+from matrix.common.dynamo_handler import OutputTableField
+from matrix.common.dynamo_utils import Lock
+
+
+ZARR_OUTPUT_CONFIG = {
+    "cells_per_chunk": 3000,
+    "compressor": zarr.storage.default_compressor,
+    "dtypes": {
+        "expression": "<f4",
+        "cell_id": "<U64",
+        "cell_metadata": "<f4"
+    },
+    "order": "C"
+}
+
+
+class S3ZarrStore:
+
+    def __init__(self, request_id: str):
+        self._request_id = request_id
+        self._results_bucket = os.environ['S3_RESULTS_BUCKET']
+        self._zarr_lock_table_name = os.environ['DYNAMO_LOCK_TABLE_NAME']
+        self._cells_per_chunk = ZARR_OUTPUT_CONFIG['cells_per_chunk']
+        self.dynamo_handler = DynamoHandler()
+        self.s3_file_system = s3fs.S3FileSystem(anon=False)
+        self.exp_df = None
+        self.qc_df = None
+
+    def write_from_pandas_dfs(self, exp_df: DataFrame, qc_df: DataFrame, num_rows: int):
+        """Write specified number of rows from matrix dataframes to s3 results bucket.
+
+        Input:
+            exp_df: (DataFrame) expression pandas dataframe from dss zarr store
+            qc_df: (DataFrame) qc values pandas dataframe from dss zarr store
+            num_rows: (int) number of rows to write from input dataframes
+        """
+        self.exp_df = exp_df
+        self.qc_df = qc_df
```

Comment:
Feels like a weird pattern. Would opt for one of the following:
- set on init
- setter method
- always pass as function params